OC-2560: Validator Lockdown

christophermaier opened the pull request:
Lockdown validators so they can only be used to create clients.
Further details in commit messages.

@doubt72, there are some tweaks to the principals endpoint I'd
appreciate a 2nd pair of eyes on.

Commit:
Add validator restrictions

A validator client should only be able to create new non-admin, non-validator clients; ALL other operations are expressely forbidden.

Commit:
Enable validator restrictions

This locks down permissions across all endpoints to restrict what validators can do.  Anything that was a blanket authorization is now denied to validators.

One exception is the principals endpoint, which does not have proper authentication enabled yet.  This endpoint is specially handled, and is currently the only one that validators can even access.  This is only a temporary solution until the principals endpoint is complete.

Commit:
Remove unused allow_validator/1 function

It wasn't used, and there's only one scenario where validators should ever be allowed anyway.

Commit:
Consolidate authz rules

Since all these functions return a simple atom, and all the logic takes place in the function heads, I've squashed all the heads onto a single line each, aligning the return values.  This makes things easier to read, and makes it easier to grasp the "big picture".

Removed Dialyzer specs that would be inferred by Dialyzer with no problems (i.e., the exact inputs and outputs are already explicitly stated in the function heads) to further reduce noise.

seth commented on src/chef_wm_base.erl:
```diff
@@ -606,10 +606,16 @@ handle_auth_info(chef_wm_clients, Req,
+
             IsValidator = chef_wm_authz:is_validator(Requestor),
-            case IsAdmin orelse (IsValidator andalso NotCreatingAdmin) of
+            CreatingUnprivileged = (ej:get({<<"admin">>}, Client) =:= false) and
```
I think admin is an optional field. But gets added during normalization so this should be OK. I suppose we might as well use `andalso`. `and` evaluates both args always.

christophermaier replied:
I switched it to `andalso`.

seth commented on src/chef_wm_base.erl:
```diff
@@ -606,10 +606,16 @@ handle_auth_info(chef_wm_clients, Req,
+            CreatingUnprivileged = (ej:get({<<"admin">>}, Client) =:= false) and
+                (ej:get({<<"validator">>}, Client) =:= false),
+
+            %% Admins can create whatever they want, but validators can only create
```
So I keep wondering if `if` helps for some of these. Just trying it out, fine as-is.

```
if
    IsAdmin -> authorized;
    (IsValidator andalso CreatingUnprivileged) -> authorized;
    true -> forbidden
end
```

christophermaier replied:
I get the desire to use `if`, but the fact that the default case has to be `true` rubs me the wrong way... it just reads oddly to me.

seth replied:
Despite having proffered the example, I actually quite agree. Disregard the if.

seth commented on src/chef_wm_base.erl:
```diff
@@ -628,9 +634,10 @@ handle_auth_info(chef_wm_named_client, Req, #base_state{requestor = Requestor,
-            #chef_client{validator = IsValidator} = Client,
-            case IsValidator of
-                true -> %% We can't delete the validator
+            case chef_wm_authz:is_validator(Client) of
```
I realize you're just cleaning this part of the code, but since we are not enforcing single validator for create AFAICT, I wonder if we should remove this check. Delete the validator if you want to.

christophermaier replied:
I'll go back through when I make validator deletion analogous to how we do admin deletion (i.e., delete all you want, unless you only have one left).  I didn't change any behavior here right now, though.

seth commented:
overall looks good to me. comments made inline.

Commit:
Laziness is a thing

doubt72 commented:
Looks okay from here.

Commit:
Delete all the clients you want

Validators are no longer privileged in this regard; you can delete 'em all if you really want to.

seth commented:
This has been hanging for awhile and is no longer an auto-merge. If you can verify that it still applies and tests pass, please MWR.
